File: apps/visualizer/backend/jobs/checkpoint.py
"""Job checkpoint helpers for persisting resume state in ``jobs.metadata[\"checkpoint\"]``.

Checkpoint payloads are versioned with ``checkpoint_version: 1`` and a fingerprint so
stale metadata from different inputs is ignored (handlers log ``checkpoint mismatch``).

**batch_describe** — ``job_type``, ``fingerprint``, ``processed_pairs`` (``\"key|itype\"`` strings),
``total_at_start``. Fingerprint includes sorted ``perspective_slugs`` when metadata carries a
non-empty list (empty/missing → ``null`` in the canonical payload).

**vision_match** — ``job_type``, ``fingerprint``, ``processed_media_keys`` (Instagram dump
``media_key`` values).

**enrich_catalog** — ``job_type``, ``fingerprint``, ``processed_image_keys`` (catalog image keys).

**prepare_catalog** — ``job_type``, ``fingerprint``, ``processed_image_keys`` (catalog image keys).

**batch_score** — ``job_type``, ``fingerprint``, ``processed_triplets`` (``\"key|itype|slug\"`` strings),
``total_at_start`` (int). Fingerprint canonical JSON matches ``batch_describe`` knobs plus an ordered
triple list (sorted by ``f\"{key}|{itype}|{slug}\"``).

**batch_analyze** — top-level ``job_type: 'batch_analyze'`` with ``stage`` (``'describe'`` or
``'score'``) and two nested objects: ``describe`` (``fingerprint``, ``processed_pairs``,
``total_at_start``) and ``score`` (``fingerprint``, ``processed_triplets``, ``total_at_start``).
``checkpoint_version: 1`` still applies at the merged checkpoint root. Fingerprints for each
``batch_analyze`` sub-object are computed exactly as ``batch_describe`` / ``batch_score`` over
their own slice of ``metadata`` (the analyze handler normalizes ``force_describe`` /
``force_score`` into each sub-payload before fingerprinting).

**batch_text_embed** — ``job_type``, ``fingerprint``, ``processed_pairs`` (``"key|catalog"`` strings),
``total_at_start`` (same semantics as ``batch_describe``).

**batch_embed_image** — ``job_type: 'batch_embed_image'``, ``fingerprint``, ``processed_pairs`` (sorted list of
catalog ``image_key`` strings — the same strings that appear in the fingerprint's ``pairs`` array),
``total_at_start``, ``checkpoint_version: 1``.

**batch_stack_detect** — ``job_type``, ``fingerprint``, ``processed_image_keys`` (catalog ``images.key``
values, sorted on persist), ``total_at_start`` (int), ``checkpoint_version: 1``. Fingerprint includes
``delta_ms`` (resolved ms), ``force_mode`` (``incremental`` | ``full`` | ``preserve_edited``), and
``keys`` (sorted work-list at job start, before checkpoint resume filtering).
"""

from __future__ import annotations

import hashlib
import json
from typing import Any

from lightroom_tagger.core.clip_embedding_service import CLIP_EMBED_DIM, CLIP_EMBED_MODEL_ID
from lightroom_tagger.core.embedding_service import TEXT_EMBED_DIM, TEXT_EMBED_MODEL_ID

CHECKPOINT_VERSION: int = 1


def fingerprint_batch_describe(
    metadata: dict[str, Any], ordered_pairs: list[tuple[str, str]]
) -> str:
    """SHA-256 hex of canonical JSON for batch_describe inputs and ordered pair list."""
    min_rating_raw = metadata.get("min_rating")
    min_rating: int | None
    try:
        min_rating = int(min_rating_raw) if min_rating_raw is not None else None
    except (TypeError, ValueError):
        min_rating = None
    pairs = [f"{key}|{itype}" for key, itype in ordered_pairs]
    ps_raw = metadata.get("perspective_slugs")
    if isinstance(ps_raw, list) and len(ps_raw) > 0:
        perspective_slugs_fp: Any = sorted(str(x) for x in ps_raw)
    else:
        perspective_slugs_fp = None
    payload = {
        "backfill_visual_tags": bool(metadata.get("backfill_visual_tags", False)),
        "date_filter": metadata.get("date_filter", "all"),
        "force": bool(metadata.get("force", False)),
        "image_type": metadata.get("image_type", "both"),
        "max_workers": int(metadata.get("max_workers", 4)),
        "min_rating": min_rating,
        "pairs": pairs,
        "perspective_slugs": perspective_slugs_fp,
        "provider_id": metadata.get("provider_id"),
        "provider_model": metadata.get("provider_model"),
    }
    canonical = json.dumps(payload, sort_keys=True, separators=(",", ":"))
    return hashlib.sha256(canonical.encode("utf-8")).hexdigest()


def fingerprint_batch_text_embed(
    metadata: dict[str, Any], ordered_pairs: list[tuple[str, str]]
) -> str:
    """SHA-256 hex of canonical JSON for batch_text_embed inputs and ordered pair list."""
    min_rating_raw = metadata.get("min_rating")
    min_rating: int | None
    try:
        min_rating = int(min_rating_raw) if min_rating_raw is not None else None
    except (TypeError, ValueError):
        min_rating = None
    pairs = sorted(f"{key}|{itype}" for key, itype in ordered_pairs)
    payload = {
        "date_filter": metadata.get("date_filter", "all"),
        "embedding_dim": TEXT_EMBED_DIM,
        "embedding_model_id": TEXT_EMBED_MODEL_ID,
        "force": bool(metadata.get("force", False)),
        "image_type": str(metadata.get("image_type", "catalog")),
        "min_rating": min_rating,
        "pairs": pairs,
    }
    canonical = json.dumps(payload, sort_keys=True, separators=(",", ":"))
    return hashlib.sha256(canonical.encode("utf-8")).hexdigest()


def _normalized_batch_embed_image_type(metadata: dict[str, Any]) -> str:
    """Canonical embed scope for resume identity: ``catalog`` vs ``catalog_and_instagram``."""
    raw = metadata.get("image_type", "catalog")
    s = str(raw).strip() if raw is not None else "catalog"
    if s == "catalog_and_instagram":
        return "catalog_and_instagram"
    return "catalog"


def fingerprint_catalog_cache_build(
    metadata: dict[str, Any],
    *,
    resolved_months: int | None = None,
    resolved_year: str | None = None,
) -> str:
    """SHA-256 hex for catalog_cache_build composite job identity (metadata knobs only).

    Mirrors embed/stack/similarity inputs that affect how much work runs — without embedding
    ordered key lists (those vary per catalog snapshot).
    """
    min_rating_raw = metadata.get("min_rating")
    min_rating: int | None
    try:
        min_rating = int(min_rating_raw) if min_rating_raw is not None else None
    except (TypeError, ValueError):
        min_rating = None
    payload: dict[str, Any] = {
        "embedding_dim": CLIP_EMBED_DIM,
        "embedding_model_id": CLIP_EMBED_MODEL_ID,
        "force_embed": bool(metadata.get("force_embed", False)),
        "force_similarity": bool(metadata.get("force_similarity", False)),
        "force_stack": bool(metadata.get("force_stack", False)),
        "image_type": "catalog_and_instagram",
        "last_months": metadata.get("last_months"),
        "min_rating": min_rating,
        "month": metadata.get("month"),
        "resolved_months": resolved_months,
        "resolved_year": resolved_year,
        "year": metadata.get("year"),
    }
    canonical = json.dumps(payload, sort_keys=True, separators=(",", ":"))
    return hashlib.sha256(canonical.encode("utf-8")).hexdigest()


def fingerprint_batch_embed_image(
    metadata: dict[str, Any],
    ordered_keys: list[str],
    *,
    resolved_months: int | None = None,
    resolved_year: str | None = None,
) -> str:
    """SHA-256 hex of canonical JSON for batch_embed_image inputs and ordered key list.

    Pass ``resolved_months`` and ``resolved_year`` from ``_resolve_date_window`` so the
    fingerprint reflects the *effective* date window rather than the raw metadata string,
    which may represent the same window under multiple spellings (``date_filter='3months'``
    vs ``last_months=3``).
    """
    min_rating_raw = metadata.get("min_rating")
    min_rating: int | None
    try:
        min_rating = int(min_rating_raw) if min_rating_raw is not None else None
    except (TypeError, ValueError):
        min_rating = None
    pairs = sorted(ordered_keys)
    payload = {
        "embedding_dim": CLIP_EMBED_DIM,
        "embedding_model_id": CLIP_EMBED_MODEL_ID,
        "force": bool(metadata.get("force", False)),
        "image_type": _normalized_batch_embed_image_type(metadata),
        "min_rating": min_rating,
        "pairs": pairs,
        "resolved_months": resolved_months,
        "resolved_year": resolved_year,
    }
    canonical = json.dumps(payload, sort_keys=True, separators=(",", ":"))
    return hashlib.sha256(canonical.encode("utf-8")).hexdigest()


def fingerprint_batch_stack_detect(
    metadata: dict[str, Any],
    image_keys: list[str],
    *,
    resolved_delta_ms: int,
    force_mode: str,
) -> str:
    """SHA-256 hex of canonical JSON for stack detect run identity (work list + delta + force).

    ``image_keys`` is the **initial** work list (entirety before resume filtering);
    ``sorted(image_keys)`` is stored in the payload as ``keys``.
    ``force_mode`` is ``incremental``, ``full``, or ``preserve_edited`` (caller must normalize per job contract).
    """
    _ = metadata  # Reserved for future fingerprint fields (e.g. filters) without breaking callers
    keys_sorted = sorted(image_keys)
    payload: dict[str, Any] = {
        "delta_ms": int(resolved_delta_ms),
        "force_mode": str(force_mode),
        "keys": keys_sorted,
    }
    canonical = json.dumps(payload, sort_keys=True, separators=(",", ":"))
    return hashlib.sha256(canonical.encode("utf-8")).hexdigest()


def fingerprint_batch_score(
    metadata: dict[str, Any], work_units: list[tuple[str, str, str]]
) -> str:
    """SHA-256 hex of canonical JSON for batch_score inputs and ordered work-unit triples."""
    min_rating_raw = metadata.get("min_rating")
    min_rating: int | None
    try:
        min_rating = int(min_rating_raw) if min_rating_raw is not None else None
    except (TypeError, ValueError):
        min_rating = None
    ordered = sorted(work_units, key=lambda u: f"{u[0]}|{u[1]}|{u[2]}")
    triples = [f"{key}|{itype}|{slug}" for key, itype, slug in ordered]
    ps_raw = metadata.get("perspective_slugs")
    if isinstance(ps_raw, list) and len(ps_raw) > 0:
        perspective_slugs_fp: Any = sorted(str(x) for x in ps_raw)
    else:
        perspective_slugs_fp = None
    payload = {
        "date_filter": metadata.get("date_filter", "all"),
        "force": bool(metadata.get("force", False)),
        "image_type": metadata.get("image_type", "both"),
        "max_workers": int(metadata.get("max_workers", 4)),
        "min_rating": min_rating,
        "perspective_slugs": perspective_slugs_fp,
        "provider_id": metadata.get("provider_id"),
        "provider_model": metadata.get("provider_model"),
        "triples": triples,
    }
    canonical = json.dumps(payload, sort_keys=True, separators=(",", ":"))
    return hashlib.sha256(canonical.encode("utf-8")).hexdigest()


def fingerprint_vision_match(
    *,
    threshold: float,
    weights: dict[str, Any],
    month: Any,
    year: Any,
    last_months: Any,
    media_key: Any,
    force_reprocess: bool,
    force_descriptions: bool,
    skip_undescribed: bool,
    provider_id: Any,
    provider_model: Any,
    max_workers: int,
    clip_top_k: int = 50,
) -> str:
    """SHA-256 hex of canonical JSON for vision_match checkpoint scope."""
    stable_weights = {k: weights[k] for k in sorted(weights)}
    payload = {
        "force_descriptions": bool(force_descriptions),
        "force_reprocess": bool(force_reprocess),
        "last_months": last_months,
        "max_workers": int(max_workers),
        "media_key": media_key,
        "month": month,
        "provider_id": provider_id,
        "provider_model": provider_model,
        "skip_undescribed": bool(skip_undescribed),
        "threshold": float(threshold),
        "weights": stable_weights,
        "year": year,
    }
    # Omit default so fingerprints match checkpoints from builds before clip_top_k existed.
    if int(clip_top_k) != 50:
        payload["clip_top_k"] = int(clip_top_k)
    canonical = json.dumps(payload, sort_keys=True, separators=(",", ":"))
    return hashlib.sha256(canonical.encode("utf-8")).hexdigest()


def fingerprint_catalog_keys(*, total: int, keys: list[str]) -> str:
    """SHA-256 hex for enrich_catalog / prepare_catalog input identity."""
    payload = {"keys": sorted(keys), "total": int(total)}
    canonical = json.dumps(payload, sort_keys=True, separators=(",", ":"))
    return hashlib.sha256(canonical.encode("utf-8")).hexdigest()


def merge_checkpoint_into_metadata(
    existing_metadata: dict[str, Any], checkpoint_body: dict[str, Any]
) -> dict[str, Any]:
    """Shallow-copy metadata and set ``checkpoint`` to a versioned body."""
    out = dict(existing_metadata)
    out["checkpoint"] = {"checkpoint_version": CHECKPOINT_VERSION, **checkpoint_body}
    return out

File: apps/visualizer/backend-ts/src/jobs/checkpoint.ts
/**
 * Resume checkpoints, stored in `jobs.metadata.checkpoint`.
 * Port of `jobs/checkpoint.py`.
 *
 * A batch job describes tens of thousands of images through a paid API, so a
 * restart must not re-run the work already paid for. The checkpoint records which
 * units are done, and a fingerprint of the inputs that produced them — resuming
 * against a *different* selection would silently skip images the user just asked
 * for, so a fingerprint mismatch discards the checkpoint and says so in the log.
 *
 * `batch_describe` — `job_type`, `fingerprint`, `processed_pairs` (`"key|itype"`),
 * `total_at_start`.
 *
 * `batch_score` — the same shape under `processed_triplets`, whose units carry a
 * perspective slug as well: `"key|itype|slug"`.
 *
 * `batch_embed_image` — the same shape, except `processed_pairs` holds bare
 * catalog keys. The key name is not a mistake to tidy up: it is what the Flask
 * backend already wrote, and renaming it here would strand every checkpoint the
 * cutover is meant to resume.
 *
 * `batch_stack_detect` — bare catalog keys again, but under
 * `processed_image_keys`. Same reason: the name is the Flask backend's.
 *
 * `batch_analyze` — the composite, and the one that is not flat: `stage` names the
 * pass that is running, and `describe` / `score` each hold a body of the shape
 * above. Two fingerprints rather than one because the stages are resumable
 * independently — a run interrupted during scoring must not re-describe.
 */
import { CLIP_EMBED_DIM, CLIP_EMBED_MODEL_ID } from '../imaging/clip-embed.js';

export const CHECKPOINT_VERSION = 1;

/** Refuse to grow a checkpoint past this; `jobs.metadata` is one JSON column. */
export const CHECKPOINT_MAX_ENTRIES = 100_000;

/**
 * `json.dumps(payload, sort_keys=True, separators=(",", ":"))`, byte for byte.
 *
 * The fingerprint has to survive the cutover: a job checkpointed by the Flask
 * backend must resume under this one rather than restarting from zero, and that
 * only holds if both sides hash the same bytes. Python sorts object keys and
 * escapes non-ASCII, neither of which `JSON.stringify` does.
 */
export function canonicalJson(value: unknown): string {
  return escapeNonAscii(JSON.stringify(sortKeys(value)));
}

function sortKeys(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(sortKeys);
  if (value === null || typeof value !== 'object') return value;
  const out: Record<string, unknown> = {};
  for (const key of Object.keys(value as Record<string, unknown>).sort()) {
    out[key] = sortKeys((value as Record<string, unknown>)[key]);
  }
  return out;
}

function escapeNonAscii(json: string): string {
  // Only string contents can hold non-ASCII, and `JSON.stringify` has already
  // escaped everything else, so a blind pass over the output is safe. Lone
  // surrogates escape individually, which is what Python emits too.
  return json.replace(/[\u0080-\uffff]/g, (c) =>
    `\\u${c.charCodeAt(0).toString(16).padStart(4, '0')}`,
  );
}

async function sha256Hex(text: string): Promise<string> {
  const digest = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(text));
  return [...new Uint8Array(digest)].map((b) => b.toString(16).padStart(2, '0')).join('');
}

/** Coerce a metadata value to an int, or `null` when it is not one. */
function asIntOrNull(raw: unknown): number | null {
  if (raw === null || raw === undefined) return null;
  const n = typeof raw === 'number' ? raw : Number(raw);
  return Number.isFinite(n) ? Math.trunc(n) : null;
}

/** Sorted `perspective_slugs`, or `null` when the list is absent or empty. */
function perspectiveSlugsFingerprint(metadata: Record<string, unknown>): string[] | null {
  const raw = metadata['perspective_slugs'];
  if (!Array.isArray(raw) || raw.length === 0) return null;
  return raw.map((x) => String(x)).sort();
}

/** Identity of a `batch_describe` run: its knobs plus the ordered work list. */
export function fingerprintBatchDescribe(
  metadata: Record<string, unknown>,
  orderedPairs: readonly (readonly [string, string])[],
): Promise<string> {
  return sha256Hex(
    canonicalJson({
      backfill_visual_tags: Boolean(metadata['backfill_visual_tags']),
      date_filter: metadata['date_filter'] ?? 'all',
      force: Boolean(metadata['force']),
      image_type: metadata['image_type'] ?? 'catalog',
      max_workers: asIntOrNull(metadata['max_workers'] ?? 4) ?? 4,
      min_rating: asIntOrNull(metadata['min_rating']),
      pairs: orderedPairs.map(([key, itype]) => `${key}|${itype}`),
      perspective_slugs: perspectiveSlugsFingerprint(metadata),
      provider_id: metadata['provider_id'] ?? null,
      provider_model: metadata['provider_model'] ?? null,
    }),
  );
}

/**
 * Identity of a `batch_score` run: its knobs plus the sorted work triples.
 *
 * The unit is `key|itype|slug`, not `key|itype`, because one image is scored once
 * per perspective and each of those is separately resumable. Activating a
 * perspective mid-run therefore changes the fingerprint, which is the point: the
 * checkpoint's "done" set says nothing about the rubric that just appeared.
 */
export function fingerprintBatchScore(
  metadata: Record<string, unknown>,
  workTriples: readonly (readonly [string, string, string])[],
): Promise<string> {
  return sha256Hex(
    canonicalJson({
      date_filter: metadata['date_filter'] ?? 'all',
      force: Boolean(metadata['force']),
      image_type: metadata['image_type'] ?? 'catalog',
      max_workers: asIntOrNull(metadata['max_workers'] ?? 4) ?? 4,
      min_rating: asIntOrNull(metadata['min_rating']),
      perspective_slugs: perspectiveSlugsFingerprint(metadata),
      provider_id: metadata['provider_id'] ?? null,
      provider_model: metadata['provider_model'] ?? null,
      triples: workTriples.map(([key, itype, slug]) => `${key}|${itype}|${slug}`).sort(),
    }),
  );
}

/**
 * Identity of a `batch_embed_image` run.
 *
 * The date window arrives already resolved rather than being re-read from the
 * metadata, because the same window has several spellings — `date_filter:
 * '3months'` and `last_months: 3` mean one thing — and fingerprinting the raw
 * text would discard a checkpoint over a change of wording.
 *
 * The model id and dimension are in the payload because a re-embed under a
 * different model is a different run: resuming across that would leave the table
 * holding two incompatible vector spaces.
 */
export function fingerprintBatchEmbedImage(
  metadata: Record<string, unknown>,
  orderedKeys: readonly string[],
  window: { resolvedMonths: number | null; resolvedYear: string | null },
): Promise<string> {
  return sha256Hex(
    canonicalJson({
      embedding_dim: CLIP_EMBED_DIM,
      embedding_model_id: CLIP_EMBED_MODEL_ID,
      force: Boolean(metadata['force']),
      image_type: 'catalog',
      min_rating: asIntOrNull(metadata['min_rating']),
      pairs: [...orderedKeys].sort(),
      resolved_months: window.resolvedMonths,
      resolved_year: window.resolvedYear,
    }),
  );
}

/**
 * Identity of a `batch_stack_detect` run: the work list, the burst gap, and the
 * force mode.
 *
 * The gap and the mode arrive resolved, not read from the metadata, because both
 * have a default the metadata does not spell — `delta_ms` falls back to
 * `config.yaml` and an absent `force` means `incremental`. A run resumed after the
 * config changed is a different grouping of the same photos, so it must not match.
 */
export function fingerprintBatchStackDetect(
  imageKeys: readonly string[],
  opts: { resolvedDeltaMs: number; forceMode: string },
): Promise<string> {
  return sha256Hex(
    canonicalJson({
      delta_ms: Math.trunc(opts.resolvedDeltaMs),
      force_mode: String(opts.forceMode),
      keys: [...imageKeys].sort(),
    }),
  );
}

/**
 * Identity of a `catalog_cache_build` run: the knobs its four stages read.
 *
 * No key list, unlike every other fingerprint here, because each stage selects
 * its own and all four move with the catalog — a chain identified by them would
 * have a different identity on every run. Nothing resumes on this: the chain
 * keeps no checkpoint, and the value exists so a completed run's result says
 * which set of knobs produced it.
 */
export function fingerprintCatalogCacheBuild(
  metadata: Record<string, unknown>,
  window: { resolvedMonths: number | null; resolvedYear: string | null },
): Promise<string> {
  return sha256Hex(
    canonicalJson({
      embedding_dim: CLIP_EMBED_DIM,
      embedding_model_id: CLIP_EMBED_MODEL_ID,
      force_embed: Boolean(metadata['force_embed']),
      force_similarity: Boolean(metadata['force_similarity']),
      force_stack: Boolean(metadata['force_stack']),
      image_type: 'catalog',
      last_months: metadata['last_months'] ?? null,
      min_rating: asIntOrNull(metadata['min_rating']),
      month: metadata['month'] ?? null,
      resolved_months: window.resolvedMonths,
      resolved_year: window.resolvedYear,
      year: metadata['year'] ?? null,
    }),
  );
}

export function buildBatchDescribeCheckpointBody(args: {
  fingerprint: string;
  processed: ReadonlySet<string>;
  totalAtStart: number;
}): Record<string, unknown> {
  return {
    job_type: 'batch_describe',
    fingerprint: args.fingerprint,
    processed_pairs: [...args.processed].sort(),
    total_at_start: args.totalAtStart,
  };
}

export function buildBatchScoreCheckpointBody(args: {
  fingerprint: string;
  processed: ReadonlySet<string>;
  totalAtStart: number;
}): Record<string, unknown> {
  return {
    job_type: 'batch_score',
    fingerprint: args.fingerprint,
    processed_triplets: [...args.processed].sort(),
    total_at_start: args.totalAtStart,
  };
}

export function buildBatchEmbedImageCheckpointBody(args: {
  fingerprint: string;
  processed: ReadonlySet<string>;
  totalAtStart: number;
}): Record<string, unknown> {
  return {
    job_type: 'batch_embed_image',
    fingerprint: args.fingerprint,
    processed_pairs: [...args.processed].sort(),
    total_at_start: args.totalAtStart,
  };
}

export function buildBatchStackDetectCheckpointBody(args: {
  fingerprint: string;
  processed: ReadonlySet<string>;
  totalAtStart: number;
}): Record<string, unknown> {
  return {
    job_type: 'batch_stack_detect',
    fingerprint: args.fingerprint,
    processed_image_keys: [...args.processed].sort(),
    total_at_start: args.totalAtStart,
  };
}

/** The two passes `batch_analyze` composes, and the two checkpoint sub-objects. */
export type AnalyzeStage = 'describe' | 'score';

export interface AnalyzeCheckpoint {
  /** The pass that was running when the checkpoint was written, if any. */
  stage: AnalyzeStage | null;
  describe: Record<string, unknown>;
  score: Record<string, unknown>;
}

const EMPTY_ANALYZE_CHECKPOINT: AnalyzeCheckpoint = { stage: null, describe: {}, score: {} };

/** The `batch_analyze` checkpoint, or empty sub-objects when there is none. */
export function readAnalyzeCheckpoint(metadata: Record<string, unknown>): AnalyzeCheckpoint {
  const body = checkpointBody(metadata, 'batch_analyze');
  if (body === null) return EMPTY_ANALYZE_CHECKPOINT;
  const stage = body['stage'];
  return {
    stage: stage === 'describe' || stage === 'score' ? stage : null,
    describe: asBody(body['describe']),
    score: asBody(body['score']),
  };
}

/**
 * Minimal runner surface `persistAnalyzeStageCheckpoint` needs, so this module
 * stays free of a dependency on the runner it is imported by.
 */
interface CheckpointWriter {
  readMetadata(jobId: string): Record<string, unknown>;
  persistCheckpoint(jobId: string, body: Record<string, unknown>): void;
}

/**
 * Write one stage's resume state into the composite checkpoint, preserving the
 * other stage's.
 *
 * Read-modify-write rather than a merge inside `persistCheckpoint`, which knows
 * nothing about stages. It is safe here because the passes are sequential: by the
 * time score is writing, describe's sub-object has stopped changing.
 */
export function persistAnalyzeStageCheckpoint(
  runner: CheckpointWriter,
  jobId: string,
  stage: AnalyzeStage,
  payload: Record<string, unknown>,
): void {
  const existing = readAnalyzeCheckpoint(runner.readMetadata(jobId));
  runner.persistCheckpoint(jobId, {
    job_type: 'batch_analyze',
    stage,
    describe: stage === 'describe' ? payload : existing.describe,
    score: stage === 'score' ? payload : existing.score,
  });
}

/** One stage's own body, the same shape its standalone job would checkpoint. */
export function buildAnalyzeStagePayload(args: {
  fingerprint: string;
  processed: ReadonlySet<string>;
  totalAtStart: number;
  resumeKey: ResumeKey;
}): Record<string, unknown> {
  return {
    fingerprint: args.fingerprint,
    [args.resumeKey]: [...args.processed].sort(),
    total_at_start: args.totalAtStart,
  };
}

/** Which array in a checkpoint body holds the processed units. */
export type ResumeKey = 'processed_pairs' | 'processed_triplets' | 'processed_image_keys';

export interface LoadResumeStateArgs {
  metadata: Record<string, unknown>;
  jobType: string;
  resumeKey: ResumeKey;
  fingerprint: string;
  /** Logged when a checkpoint exists but was built from different inputs. */
  mismatchMessage: string | null;
  log: (message: string) => void;
  /**
   * Read from this sub-object of a `batch_analyze` checkpoint instead of the
   * root, for a pass running as a stage rather than as its own job.
   */
  analyzeStage?: AnalyzeStage;
}

/**
 * The processed-unit set to resume from, or an empty set.
 *
 * Silent on a missing or stale-version checkpoint — there is nothing to tell the
 * user. Loud on a fingerprint mismatch, because that one means work they expected
 * to be skipped is about to be redone.
 */
export function loadResumeState(args: LoadResumeStateArgs): Set<string> {
  const body =
    args.analyzeStage === undefined
      ? checkpointBody(args.metadata, args.jobType)
      : readAnalyzeCheckpoint(args.metadata)[args.analyzeStage];
  if (body === null) return new Set();

  if (body['fingerprint'] === args.fingerprint) {
    const units = body[args.resumeKey];
    return new Set(Array.isArray(units) ? units.map((u) => String(u)) : []);
  }
  if (body['fingerprint'] && args.mismatchMessage) args.log(args.mismatchMessage);
  return new Set();
}

/** A current-version checkpoint of `jobType`, or `null`. */
function checkpointBody(
  metadata: Record<string, unknown>,
  jobType: string,
): Record<string, unknown> | null {
  const chk = metadata['checkpoint'];
  if (!chk || typeof chk !== 'object' || Array.isArray(chk)) return null;
  const body = chk as Record<string, unknown>;
  if (body['checkpoint_version'] !== CHECKPOINT_VERSION) return null;
  if (body['job_type'] !== jobType) return null;
  return body;
}

function asBody(value: unknown): Record<string, unknown> {
  return value && typeof value === 'object' && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : {};
}
